solr/contrib/ltr/src/java/org/apache/solr/ltr/feature/PrefetchingFieldValueFeature.java

```diff
@@ -0,0 +1,134 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.ltr.feature;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.solr.common.util.StrUtils;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.search.SolrDocumentFetcher;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This feature returns the value of a field in the current document.
+ * The field must have stored="true" or docValues="true" properties.
+ * Example configuration:
+ * <pre>{
+  "name":  "rawHits",
+  "class": "org.apache.solr.ltr.feature.PrefetchingFieldValueFeature",
+  "params": {
+      "field": "hits"
+  }
+}</pre>
+ */
+public class PrefetchingFieldValueFeature extends FieldValueFeature {
+  // used to store all fields from all PrefetchingFieldValueFeatures
+  private Set<String> prefetchFields;
+  // can be used for debugging to only fetch the field this features uses
+  public static final String DISABLE_PREFETCHING_FIELD_VALUE_FEATURE = "disablePrefetchingFieldValueFeature";
+
+  public void setPrefetchFields(Set<String> fields) {
+    prefetchFields = fields;
+  }
+
+  @VisibleForTesting
+  public Set<String> getPrefetchFields(){
+    return prefetchFields;
+  }
+
+  public PrefetchingFieldValueFeature(String name, Map<String,Object> params) {
+    super(name, params);
+  }
+
+  @Override
+  public FeatureWeight createWeight(IndexSearcher searcher, boolean needsScores,
+      SolrQueryRequest request, Query originalQuery, Map<String,String[]> efi)
+          throws IOException {
+    return new PrefetchingFieldValueFeatureWeight(searcher, request, originalQuery, efi);
+  }
+
+  public class PrefetchingFieldValueFeatureWeight extends FieldValueFeatureWeight {
+    private final SolrDocumentFetcher docFetcher;
+    private final Boolean disablePrefetching;
+
+    public PrefetchingFieldValueFeatureWeight(IndexSearcher searcher,
+        SolrQueryRequest request, Query originalQuery, Map<String,String[]> efi) {
+      super(searcher, request, originalQuery, efi);
+
+      disablePrefetching = request.getParams().getBool(DISABLE_PREFETCHING_FIELD_VALUE_FEATURE, false);
```

Comment:
Hmm, interesting idea, I'd like to go away and ponder it a little. My initial thoughts are that yes debugging can be facilitated this way but at a cost of (marginal) additional code complexity and from a general user perspective it's another flag to understand and use (or not use), and with that in mind my intuitive preference would be to not have it, but I'd like to go and think about it some more properly.

Reply:
Hi @cpoerschke ,
after thinking about your suggestion (in the comment below this)  for quite some time, I am still not sure what to make from it.
To make my thoughts transparent:
* yes, the parameter and the switch add some complexity to the code
   * however, the default user will most likely just stumble upon the param in the exception for the PFVF
   * in the exception it is explained what to use the param for
* the "fallback-fetch" also adds quite some complexity to `score()` because of the try-catch-chain which I find harder to read than the current status
  * downside of status quo: although being less complex inside `score()`, the parameter distributes its complexity through the whole feature all the way to the query
* I like about your idea, that one problematic field does not break all other prefetchFields
  * this happens in the status quo
  * however, this fallback means that if one field is broken, we basically have no benefit from the PFVF anymore. If users do only look at errors, they wouldn't even realize it (Sure, you shouldn't do that, but I cannot say how users handle their logs. If an exception is thrown, they at least are quickly aware of the problem)
* it bugs me that each PFVFS would try to fetch the broken field (all fields) before falling back to only getting its field. This does not feel efficient
* > If we had a fetchDocument method here for this block of code **and did not provide** a disablePrefetching option then the fetchDocument method would provide a hook
  * I cannot follow this comment of yours. Couldn't a subclass just ignore the `disablePrefetching` when implementing an own `score()` or `fetchDocument()`? (own `fetchDocument()` method was added in 8c5071b )

**The missing piece for me is the following:**
**How likely is a failing fetch?** Can it just occur sometimes when some internal problems occur (A) ? Or will it happen all the time for one field that is somehow corrupted (B)?
While looking into possible tests for the fallback I tried to get to the core of the IOException. It seemed to be quite deep down (I remember some Byte-Buffers). So I would assume, that A would apply instead of B.
In that case, this argument from me 
> if one field is broken, we basically have no benefit from the PFVF anymore

would not apply anymore, because one field would not constantly be broken.

These are my thoughts about this topic. I see arguments for both approaches and no showstoppers for either one, that's why I couldn't decide what to do.
Do you maybe have some insights about the likelihood of a fetch?
Or just general thoughts about it? Maybe I made some assumptions as base for my thoughts which aren't correct?

Reply:
I just answered your comment below about extracting `fetchDocument()` and I think that I did not understand your comment regarding the `disablePrefetchign` correctly until now. Referencing this one:
> If we had a fetchDocument method here for this block of code and did not provide a disablePrefetching option then the fetchDocument method would provide a hook

Your intention is something like this 
```java
public static final String DISABLE_PREFETCHING_FIELDS = "disablePrefetchingFieldValueFeatureFields";
private final List<String> disablePrefetching;
disablePrefetching = (List<String>) request.getParams().get(DISABLE_PREFETCHING_FIELDS, Collections.emptyList());
```
And that's why you want to remove the parameter from the PFVF because it does not have to be a boolean is all cases, right?

Reply:
Hi @tomglk,

All good points here re: the benefits of a disable prefetching mechanism and the disadvantages of try-catch-retry logic. Reading your arguments here (thanks for the details and structure!) and the configured/computed/effective work on https://github.com/cpoerschke/solr/commit/df2864a95e1a83ee446781042fe119981551a932 generated some more insights and ideas, I will try to organise them into logical chunks below but in short: I agree and can see now that and how a prefetch disabling mechanism will be useful.

> ... Your intention is something like this (in a subclass of PFVF) ...

Yes, "in a subclass of PFVF" was the intention and then the details of the subclass could suit the particular use case i.e. could be boolean or could be non-boolean in some cases. But (spoiler alert) re: the next 'logical chunk' updates below, it seems to me now that there is a general (non-subclass) case for a non-boolean parameter.

Reply:
Here's the 'logical chunks' mentioned above:

* index corruption is (hopefully) a rare source of errors, and in such a scenario rethrowing the exception with the details would seem most useful.

* missing-ness is a probably more likely potential source of errors: how would the document fetcher behave if one of the fields does not exist in the schema or it exists but is not stored and not docValues? or if the field exists but is optional and the document being scored has no value for the field? perhaps the test coverage already has the answers to this, i haven't looked, just thinking out aloud here.

* next then moving on to potential sources of missing-ness:
  * `[effective]prefetchFields` is computed from the PrefetchingFieldValueFeature itself and its fellow PrefetchingFieldValueFeature in the same feature store
  * _if_ a feature is removed from the store and _if_ that did not propagate the new joint prefetchFields to the features remaining in the store then that would potentially be inefficient i.e. a field is unnecessarily prefetched but it would not be a problem as such since the prefetch is for the underlying field rather than the removed feature
    * aside: from what i recall we don't support removal of features from a store
    * aside: re-computation of the joint prefetchFields happens when a feature is uploaded to a feature store, the upload could be for an established actively used feature store. concurrent read-use of `PrefetchingFieldValueFeature.prefetchFields` is not a problem but one thread reading and another updating seems potentially problematic

* when a feature is added to the store, at that point missing-ness could arise e.g. due to a typo in the field name or due to destination mistakes i.e. upload to storeA for collectionA was intended but upload to storeB for collectionB happened
  * if/since removal of features is not supported there is no obviously quick remedy here (probably deleting the feature store and re-uploading it with only the "good" features should work though I'm not entirely sure how any models using the store's "good" features would behave in that scenario, it probably would work whilst the models are loaded in memory but would not work across a reload or restart)
  * if a PrefetchingFieldValueFeature with a field named "bad1" was uploaded then operationally a precise "oops, sorry, please ignore that 'bad1' field" mechanism would be more targetted than a broader "ouch, let's temporarily turn off all prefetching" mechanism
  * the exception thrown could guide towards a (say) `ltr.PrefetchingFieldValueFeature.exclusions` string parameter (instead of the `disablePrefetchingFieldValueFeature` boolean parameter) and iteratively the user can then reduce `ltr.PrefetchingFieldValueFeature.exclusions=bad1,bar,foo,good1` down to `ltr.PrefetchingFieldValueFeature.exclusions=bad1` meaning that the features using the `bar,foo,good1` fields continue to benefit from prefetching
    * aside: when shrinking the `[effective]prefetchFields` based on `ltr.PrefetchingFieldValueFeature.exclusions=bad1,bar,foo,good1` the implementation should not exclude itself e.g. features using the `good1` field must disregard the `good1` element on the `ltr.PrefetchingFieldValueFeature.exclusions` list

* Might `ltr.PrefetchingFieldValueFeature.exclusions=aaa,bbb,ccc,ddd,eee,fff` have other uses too?
  * The effects of a feature store update could be somewhere in between 'good' and 'bad' e.g.
    * `model1` uses features `aaa,bbb,ccc`
    * we add `fl=[fv]` to extract feature values for old and new features in order to build `model2`
    * `fl=[fv]` was previously fine but with the new features latency is more problematic
    * we complete feature extraction and stop using `fl=[fv]`
    * `model1` does not use features `ddd,eee,fff` but since all features are in the same store `model1`'s features will prefetch fields for `ddd,eee,fff` even if `fl=[fv]` is not requested
      * `ltr.PrefetchingFieldValueFeature.exclusions=ddd,eee,fff` can eliminate the prefetch
    * we deploy `model2` and if it only uses the `ddd,eee,fff` features then `ltr.PrefetchingFieldValueFeature.exclusions=aaa,bbb,ccc` can eliminate unnecessary prefetching of the old features' fields

(I don't know if practically additional prefetch fields could work out as costly enough to justify the operational efforts of the outlined `ltr.PrefetchingFieldValueFeature.exclusions` parameter use but in principle the parameter could be used in that way.)

----

Bit of a long write-up that, sorry, did it sort of get the ideas across or maybe some bits could be clarified somehow, what do you think?

Oh and of course `ltr.PrefetchingFieldValueFeature.exclusions` is just a tentative name i.e. we can call it something else also and `ltr.PrefetchingFieldValueFeature.exclusions=*` equivalent to `disablePrefetchingFieldValueFeature=true` may also be convenient.

Reply:
Hi @cpoerschke ,

please do not worry about the length of your responses! I am extremely happy and thankful, that you take the time to share your thoughts and ideas in so much detail and with examples. :)
Besides being useful to understand you, it also sparks ideas in my brain and points out aspects that are not on my radar. Because of that collaborating with you is an amazing learning experience!

I stumbled upon this question of yours:
> how would the document fetcher behave if one of the fields does not exist in the schema or it exists but is not stored and not docValues? or if the field exists but is optional and the document being scored has no value for the field?

and added a test for it ( 4d84293 ).
No matter if a field is not existing, only indexed or empty, we always return the default.
The reason for this is that Lucene iterates over present fields and check which ones are needed, instead of trying to read all wanted fields (https://github.com/apache/lucene/blob/be94a667f2091b2c0fad570f9e726197d466767b/lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/Lucene90CompressingStoredFieldsReader.java#L653 and https://github.com/apache/lucene/blob/be94a667f2091b2c0fad570f9e726197d466767b/lucene/core/src/java/org/apache/lucene/document/DocumentStoredFieldVisitor.java#L101).

Reply:
Reading through your commit, I am wondering why you split up the `computedPrefetchFields` and the `effectivePrefetchFields`.
I do understand the difference between them, but I am not sure what to use them for apart from the output of `paramsToMap()`.
Speaking of the output of paramsToMap, I remember this input of yours
> if our two hypethical nodes both complained about some features but one complained about fZ, fA, fX, fB and the other about fA, fZ, fB, fX then it might be tricky to interpret that they are both complaining about the same thing

I think in terms of debugging, your suggestion to use a sorted set, combined with keeping the field of the feature inside the params.prefetchFields makes the most sense. 
This would result in duplicated information in the output, but I imagine the user doing stuff like "highlight" => "Strg + F" or searching the logs like "prefetchFields=fA, fB, fX, fZ" which would be impossible if we always remove the `field` from the output.

Reply:
Looking at `configuredPrefetchFields` maybe this could be an optional parameter in the features.json? 
Behaving like `ltr.PrefetchingFieldValueFeature`.~~exclusions~~`.inclusions` on a config level instead of as a query parameter.

Reply:
Regarding `ltr.PrefetchingFieldValueFeature.exclusions`:
> I don't know if practically additional prefetch fields could work out as costly enough to justify the operational efforts of the outlined ltr.PrefetchingFieldValueFeature.exclusions parameter use

I was surprised by your point that multiple models may be using the same store. I just assumed that people would use one feature-store per model.
This could really lead to a lot of unnecessary fetching, especially is there is not much overlap between the features of the models.
**However**, after taking a step back and writing the test ( 4d84293 ), this `costly enough to justify` thought of you stayed in my mind.
The exclusion is not necessary to handle problems with only indexed or not existing fields.
Also, to put it very bluntly, I am wondering how much of user's misconfiguration we should try to handle in our code.
If there are two models that barely use the same features, which really would result in a lot of unnecessary fetching to the point where performance worsens, I would expect the user to use a separate feature-store per model.
Maybe his could even be a hint / recommendation in the documentation.

This statement from you:
> index corruption is (hopefully) a rare source of errors, and in such a scenario rethrowing the exception with the details would seem most useful.

made me want to remove the `disablePrefetchingFieldValueFeature` completely. While implementing it I felt like there would be more value to it, but now it seems like this param would only be useful in a situation where something is really wrong with the index (assuming that 'useful' means that there are quite some `Unable to extract feature...`-exceptions). 
In that case the failing fetch of the PFVF seems to be the smallest problem besides like being unable to properly search anymore.

So I am thinking about:
```java
protected Document fetchDocument() throws FeatureException {
  try { 
    return docFetcher.doc(context.docBase + itr.docID(), prefetchFields);
  } catch (final IOException e) {
    final String prefetchedFields = disablePrefetching ? getField() : StrUtils.join(prefetchFields, ',');
    throw new FeatureException(...);
  }
}
```
What do you think about the above change?
We could also provide some fail-safety with your suggested try-catch. (Good structure in the method / comments could make up for the complexity. And maybe log an error instead of a warning?)

Reply:
About: 
> one thread reading and another updating seems potentially problematic

Thank you for having thread-safety / concurrency in mind! This is a topic about which I still have a lot to learn and build an awareness for.
To separate the different construction sites we are on, I would like to move the conversation about that to your comment at the bottom. (https://github.com/apache/solr/pull/166#discussion_r650491855)